src/corelibs/U2Formats/src/NewickFormat.cpp

```diff
@@ -32,70 +33,50 @@
     supportedObjectTypes += GObjectTypes::PHYLOGENETIC_TREE;
 }
 
-#define BUFF_SIZE 1024
-
-static QList<GObject *> parseTrees(IOAdapter *io, const U2DbiRef &dbiRef, const QVariantMap &fs, U2OpStatus &si);
+static QList<GObject *> parseTrees(IOAdapterReader &reader, const U2DbiRef &dbiRef, const QVariantMap &fs, U2OpStatus &si);
 
-Document *NewickFormat::loadTextDocument(IOAdapter *io, const U2DbiRef &dbiRef, const QVariantMap &fs, U2OpStatus &os) {
-    QList<GObject *> objects = parseTrees(io, dbiRef, fs, os);
+Document *NewickFormat::loadTextDocument(IOAdapterReader &reader, const U2DbiRef &dbiRef, const QVariantMap &hints, U2OpStatus &os) {
+    QList<GObject *> objects = parseTrees(reader, dbiRef, hints, os);
     CHECK_OP_EXT(os, qDeleteAll(objects), nullptr);
-    Document *d = new Document(this, io->getFactory(), io->getURL(), dbiRef, objects, fs);
-    return d;
+    return new Document(this, reader.getFactory(), reader.getURL(), dbiRef, objects, hints);
 }
 
-void NewickFormat::storeDocument(Document *d, IOAdapter *io, U2OpStatus &os) {
-    Q_UNUSED(os);
-    assert(d->getDocumentFormat() == this);
-
-    foreach (GObject *obj, d->getObjects()) {
-        PhyTreeObject *phyObj = qobject_cast<PhyTreeObject *>(obj);
-        if (phyObj != nullptr) {
-            QByteArray data = NewickPhyTreeSerializer::serialize(phyObj->getTree());
-            io->writeBlock(data.constData(), data.size());
+void NewickFormat::storeTextDocument(IOAdapterWriter &writer, Document *document, U2OpStatus &os) {
+    const QList<GObject *> &objects = document->getObjects();
+    for (GObject *obj : qAsConst(objects)) {
+        if (auto phyObj = qobject_cast<PhyTreeObject *>(obj)) {
+            QString text = NewickPhyTreeSerializer::serialize(phyObj->getTree());
+            writer.write(os, text);
+            CHECK_OP(os, );
         }
     }
 }
 
-FormatCheckResult NewickFormat::checkRawTextData(const QByteArray &rawData, const GUrl &) const {
-    const char *data = rawData.constData();
-    int size = rawData.size();
-    bool containsBinary = TextUtils::contains(TextUtils::BINARY, data, size);
-    if (containsBinary) {
-        return FormatDetection_NotMatched;
-    }
+FormatCheckResult NewickFormat::checkRawTextData(const QString &dataPrefix, const GUrl &) const {
     int brackets = 0;
     typedef enum { letter,
                    letter_than_whites,
                    any } Cases;
     Cases last = any;
     bool quotedLabelStarted = false;
-    for (int i = 0; i < size; ++i) {
-        if ('\'' == data[i]) {
-            if (!quotedLabelStarted && i > 0 && (data[i - 1] == '(' || data[i - 1] == ',')) {
-                quotedLabelStarted = true;
-            } else if (quotedLabelStarted) {
-                quotedLabelStarted = false;
-            }
+    for (int i = 0; i < dataPrefix.size(); ++i) {
+        char ch = dataPrefix[i].toLatin1();
+        if (ch == '\'') {
+            char prevCh = dataPrefix[i - 1].toLatin1();
```

Comment:
You removed check that `i` is greater than zero.

Reply:
This was good. Added a dedicated test that will not allow to happen this again.